File: leetcode/java/src/main/java/P3.java
import java.util.HashSet;
import java.util.Set;

/**
 * Created by taihejin on 16-5-29.
 */
public class P3 {

    private int[] flag = new int[256];

    private void reset() {
        for (int i = 0; i < flag.length; i++) {
            flag[i] = 0;
        }
    }

    public int lengthOfLongestSubstring(String s) {
        if (s.length() <= 1) return s.length();
        reset();
        int i = 0, j = 0, maxLength = 1;
        for (i = 0; i < s.length(); i++) {
            for (j = i; j < s.length(); j++) {
                int index = (int)s.charAt(j);
                if (flag[index] == 0) {
                    flag[index] = 1;
                } else {
                    maxLength = j - i > maxLength ? j - i : maxLength;
                    reset();
                    break;
                }
            }
            if (j == s.length()) {
                break;
            }
        }
        return j - i > maxLength ? j - i : maxLength;
    }

    private void doTest(String value) {
        System.out.println(String.format("length=%d, max substring=%d", value.length(), lengthOfLongestSubstring(value)));
    }

    private void test() {
        doTest("dvdf");
        doTest("abcabcbb");
        doTest("bbbbb");
        doTest("pwwkew");
        doTest("xbawtvebluuagttbeqbihnlucpmg");
    }

    public static void main(String[] args) {
        new P3().test();
    }

}

File: leetcode/cpp/p3.cc
#include <iostream>
#include <ctype.h>
#include <assert.h>
#include <string.h>
class Solution {
  public:
    int lengthOfLongestSubstring(std::string s) {
      int length = s.size();
      if (length <= 1) return length;
      unsigned char flag[256];
      memset(flag, 0, sizeof(flag));
      int i = 0, j = 0, maxLength = 1;
      for (; i < length; i++) {
        for (j = i; j < length; j++) {
          char ch = s[j];
          if (flag[ch] == 0) {
            flag[ch] = 1;
          } else {
            maxLength = j - i > maxLength ? j - i : maxLength;
            memset(flag, 0, sizeof(flag));
            break;
          }
        }
        if (j == length) break;
      }
      return j - i > maxLength ? j - i : maxLength;
    }
    
};

void test2() {
  Solution solution;
  std::string str0("xbawtvebluuagttbeqbihnlucpmg");
  assert(solution.lengthOfLongestSubstring(str0) == 12);
  std::string str("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~ ");
  std::cout << solution.lengthOfLongestSubstring(str) << " " << str.size();
  std::string str1("abcabcbb");
  assert(solution.lengthOfLongestSubstring(str1) == 3);
  std::string str2("bbbbb");
  assert(solution.lengthOfLongestSubstring(str2) == 1);
  std::string str3("pwwkew");
  assert(solution.lengthOfLongestSubstring(str3) == 3);
  std::string str4;
  assert(solution.lengthOfLongestSubstring(str4) == 0);
}

int main(int argc, char* argv[]) {
  test2();
  std::string str;
  std::cin >> str;
  Solution solution;
  std::cout << solution.lengthOfLongestSubstring(str);
  return 0;
}
